File: Trees/Binary Search Trees/Validate_BST.cpp
// VALIDATE BINARY SEARCH TREE --->> LEETCODE



// Given the root of a binary tree, determine if it is a valid binary search tree (BST).

// A valid BST is defined as follows:

// The left 
// subtree
//  of a node contains only nodes with keys less than the node's key.
// The right subtree of a node contains only nodes with keys greater than the node's key.
// Both the left and right subtrees must also be binary search trees.

// ALGORITHM-->
// Follow these steps while the current node is not null:
//    Process the current node and go to its right child if it doesn't have a left child.
//   Find the inorder predecessor of the current node—that is, the rightmost node in the left subtree—if the present node has a left child, and see if its value is smaller than the value of the current node.

// If the predecessor's right child is null, go to the current node's left child and change the predecessor's right child to point to it.
//  In order to restore the binary tree's original structure, reset the predecessor's right child to null, process the current node, and then move to its right child if it is already referring to the current node.





// C++ program to check if a given tree is BST.
#include <bits/stdc++.h>
using namespace std;



struct Node {
	// structure of a node of the tree.
	int data;
	struct Node *left, *right;

	Node(int data)
	{
		this->data = data;
		left = right = NULL;
	}
};


bool validate(Node* root,long long int min , long long int max){
    if(!root)
        return true;  // if the root is null then it is a valid BST. it means that the tree is empty or we had reached the end of tree. 
    // initializing the ans variable to false (temporarily).
	bool ans = false;    

	// checking if the root's data is in the range of min and max.
    if(root->data<max && root->data>min)
        ans = true;
    else    
        return ans; // if the root's data is not in the range of min and max then it is not a valid BST. hence returning false.
    
	// changing min and max for the left and right subtree. and checking for the left and right subtree with respesct to tree root and returning the ans.
	return ans && validate(root->left,min,root->data) &&
    validate(root->right,root->data,max);       
}


    bool isValidBST(Node* root) {
        if(!root)
            return true;

	// calling  validate function so that it can check for the left and right subtree .. while giving the range of the values of the nodes.
        return  validate(root ,-9223372036854775807,9223372036854775807 );
    }

int main()
{

	// Initializing the tree.
	struct Node* root = new Node(3);
	root->left = new Node(2);
	root->right = new Node(5);
	root->left->left = new Node(1);
	root->left->right = new Node(4);

	// calling the function to check BST.
	if (isValidBST(root))
		cout << "Is BST";
	else
		cout << "Not a BST";

	return 0;
}




// T.C.  O(N)
// S.C.  O(N) ---> for Auxillary stack

File: Trees/Binary Search Trees/Validate_BST.js
/*
	Write a function that takes in a potentially invalid Binary Search Tree (BST)
  	and returns a boolean representing whether the BST is valid.
  	Sample Input :
   	     10
       /     \
      5      15
    /   \   /   \
   2     5 13   22
 /           \
1            14
	Output : True

	Explanation:
	This code defines a Binary Search Tree (BST) struct with an integer value and left and right nodes that can 
	point to other BST nodes. The struct also has a method called ValidateBst() that returns a boolean indicating 
	whether the tree is a valid BST or not.

	The BST struct has another method called validateBST() that is used by ValidateBst() to check whether the tree
	is a valid BST or not. The validateBST() method takes in two arguments, min and max, which represent the minimum 
	and maximum values that the current node's value can take in order to be a valid BST.

	The validateBST() method first checks whether the current node's value is within the valid range determined 
	by the min and max arguments. If not, the method returns false, indicating that the tree is not a valid BST.

	If the current node's value is within the valid range, the method then recursively calls itself on the left 
	and right child nodes to check whether their values are within their valid ranges. The valid range for the 
	left child node is defined by the minimum value and the parent node's value, while the valid range for the 
	right child node is defined by the parent node's value and the maximum value.

	If all of the nodes in the tree satisfy the BST property, the method returns true, indicating that the tree 
	is a valid BST.

	O(n) time | O(d) space - where n is the number of nodes in the BST and d is the depth (height) of the BST

*/
class Node {
     constructor(data) {
         this.data = data;
         this.left = null;
         this.right = null;
     }
 }
 
// isBst is a method of BST that checks if the binary search tree is valid 
function isBST(root) {
     let prev = null;
 
     function inorderTraversal(node) {

        // Base case
         if (node === null) {
             return true;
         }
         
         // recursively check the left subtree, making sure all values are less than the current node's value
         if (!inorderTraversal(node.left)) {
             return false;
         }
 
        // if the current node's value is outside the allowed range, then the tree is invalid
         if (prev !== null && node.data <= prev.data) {
             return false;
         }
 
         prev = node;
         return inorderTraversal(node.right);
     }
 
     return inorderTraversal(root);
 }
 
 // Create the tree
 const root = new Node(100);
 root.left = new Node(-2);
 root.right = new Node(3);
 root.left.left = new Node(-4);
 root.left.right = new Node(-1);
 root.right.left = new Node(2);
 root.right.right = new Node(7);
 
 // Function call
 if (isBST(root)) {
     console.log("Is Valid BST");
 } else {
     console.log("Not a Valid BST");
 }
 

////// Complexity:
// Time  Complexity: O(n) where n is number of nodes
// Space Complexity: O(1)
